File: app/src/main/java/com/example/cameraaccess/upload/repository/VideoUploadRepository.kt
package com.example.cameraaccess.upload.repository

import android.util.Log
import com.example.cameraaccess.data.model.ApiCache
import com.example.cameraaccess.data.model.SiteBlockCreateResponse
import com.google.gson.Gson
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.RequestBody.Companion.asRequestBody
import java.io.File
import java.io.RandomAccessFile
import kotlin.math.ceil
import kotlinx.coroutines.ensureActive
import kotlin.coroutines.coroutineContext

class VideoUploadRepository(
    private val apiCache: ApiCache
) {
    private val okHttpClient: OkHttpClient = OkHttpClient()

    suspend fun uploadVideoInChunks(
        videoPath: String,
        presignedUrls: List<String>,
        chunkSize: Long = 500L * 1024 * 1024, // 500MB
        onProgress: (Float) -> Unit,
        isCancelled: () -> Boolean
    ): List<String> {

        val file = File(videoPath)
        val fileSize = file.length()
        val totalParts = ceil(fileSize / chunkSize.toDouble()).toInt()

        val raf = RandomAccessFile(file, "r")
        val eTags = mutableListOf<String>()

        var uploadedBytes = 0L

        for (partIndex in 0 until totalParts) {
            coroutineContext.ensureActive()
            if (isCancelled()) break

            val offset = partIndex * chunkSize
            val remaining = fileSize - offset
            val currentChunkSize = minOf(chunkSize, remaining)

            val buffer = ByteArray(currentChunkSize.toInt())
            raf.seek(offset)
            raf.readFully(buffer)

            val requestBody = buffer.toRequestBody("video/mp4".toMediaType())

            val url = presignedUrls[partIndex]
            Log.d("UPLOAD", "Uploading chunk ${partIndex + 1}/$totalParts to $url")

            val request = Request.Builder()
                .url(url)
                .put(requestBody)
                .build()

            val response = okHttpClient.newCall(request).execute()

            val responseBody = response.body?.string()
            Log.d(
                "UPLOAD",
                "Chunk ${partIndex + 1} response: code=${response.code}, headers=${response.headers}"
            )

            if (!response.isSuccessful) {
                Log.e(
                    "UPLOAD",
                    "Chunk ${partIndex + 1} FAILED: code=${response.code}, body=$responseBody"
                )
                throw Exception(
                    "Chunk ${partIndex + 1} failed: HTTP ${response.code}, body=$responseBody"
                )
            }

            val etag = response.header("ETag") ?: ""
            eTags.add(etag)

            uploadedBytes += currentChunkSize
            onProgress(uploadedBytes.toFloat() / fileSize.toFloat())
        }

        raf.close()
        return eTags
    }

    fun uploadCsvFile(
        csvPath: String,
        presignedUrl: String
    ) {
        val file = File(csvPath)
        val body = file.asRequestBody("text/csv".toMediaType())

        val request = Request.Builder()
            .url(presignedUrl)
            .put(body)
            .build()

        val response = okHttpClient.newCall(request).execute()
        val responseBody = response.body?.string()

        Log.d(
            "UPLOAD",
            "CSV upload response: code=${response.code}, headers=${response.headers}"
        )

        if (!response.isSuccessful) {
            Log.e(
                "UPLOAD",
                "CSV upload FAILED: code=${response.code}, body=$responseBody"
            )
            throw Exception("CSV upload failed: HTTP ${response.code}, body=$responseBody")
        }
    }

    fun uploadGpxFile(
        gpxPath: String,
        presignedUrl: String
    ) {
        val file = File(gpxPath)
        val body = file.asRequestBody("application/gpx+xml".toMediaType())

        val request = Request.Builder()
            .url(presignedUrl)
            .put(body)
            .build()

        val response = okHttpClient.newCall(request).execute()
        val responseBody = response.body?.string()

        Log.d(
            "UPLOAD",
            "GPX upload response: code=${response.code}, headers=${response.headers}"
        )

        if (!response.isSuccessful) {
            Log.e(
                "UPLOAD",
                "GPX upload FAILED: code=${response.code}, body=$responseBody"
            )
            throw Exception("GPX upload failed: HTTP ${response.code}, body=$responseBody")
        }
    }

    fun getPresignedVideoUrls(): List<String> {
        val submitResponseJson = apiCache.getSubmitSiteBlock() ?: return emptyList()
        val submitResponse =
            Gson().fromJson(submitResponseJson, SiteBlockCreateResponse::class.java)

        return submitResponse.videoPresignedUrl?.let { listOf(it) } ?: emptyList()
    }

    fun getPresignedGpxUrl(): String? {
        val submitResponseJson = apiCache.getSubmitSiteBlock() ?: return null
        val submitResponse =
            Gson().fromJson(submitResponseJson, SiteBlockCreateResponse::class.java)

        return submitResponse.gpxPresignedUrl
    }

    fun getPresignedCsvUrl(): String? {
        val submitResponseJson = apiCache.getSubmitSiteBlock() ?: return null
        val submitResponse =
            Gson().fromJson(submitResponseJson, SiteBlockCreateResponse::class.java)

        return submitResponse.csvPresignedUrl
    }
}

File: TauResearch/Repositories/VideoUploadRepository.swift
// VideoUploadRepository.swift
// Mirrors upload/repository/VideoUploadRepository.kt — chunked PUTs to presigned S3 URLs.

import Foundation
import os.log

struct VideoUploadRepository {

    private let log = Logger(subsystem: "com.tau.research", category: "upload")
    private let session: URLSession = .shared

    // 500 MB — matches Android implementation.
    static let chunkSize: Int64 = 500 * 1024 * 1024

    // MARK: - Video (chunked PUT)

    func uploadVideoInChunks(
        videoPath: String,
        presignedUrls: [String],
        progress: @escaping (Float) -> Void,
        isCancelled: () -> Bool
    ) async throws -> [String] {
        let fileURL = URL(fileURLWithPath: videoPath)
        let handle = try FileHandle(forReadingFrom: fileURL)
        defer { try? handle.close() }

        let fileSize = Int64(
            (try? fileURL.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0
        )
        let totalParts = Int((Double(fileSize) / Double(Self.chunkSize)).rounded(.up))
        var etags: [String] = []
        var uploadedBytes: Int64 = 0

        for partIndex in 0..<totalParts {
            if isCancelled() { break }
            let offset = Int64(partIndex) * Self.chunkSize
            let remaining = fileSize - offset
            let thisChunk = min(Self.chunkSize, remaining)

            try handle.seek(toOffset: UInt64(offset))
            let buffer = try handle.read(upToCount: Int(thisChunk)) ?? Data()
            guard partIndex < presignedUrls.count,
                  let url = URL(string: presignedUrls[partIndex]) else {
                throw AppError.unexpected(underlying: nil)
            }

            log.debug("Uploading chunk \(partIndex + 1)/\(totalParts, privacy: .public)")

            var req = URLRequest(url: url)
            req.httpMethod = "PUT"
            req.setValue("video/mp4", forHTTPHeaderField: "Content-Type")

            let (_, response) = try await session.upload(for: req, from: buffer)
            guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
                let code = (response as? HTTPURLResponse)?.statusCode ?? 0
                throw AppError.server(code: code, message: "Chunk \(partIndex + 1) failed")
            }
            let etag = http.value(forHTTPHeaderField: "ETag") ?? ""
            etags.append(etag)

            uploadedBytes += thisChunk
            progress(Float(uploadedBytes) / Float(fileSize))
        }
        return etags
    }

    // MARK: - GPX / CSV single PUT

    func uploadGpxFile(path: String, presignedUrl: String) async throws {
        try await uploadSingleFile(path: path, presignedUrl: presignedUrl, contentType: "application/gpx+xml")
    }

    func uploadCsvFile(path: String, presignedUrl: String) async throws {
        try await uploadSingleFile(path: path, presignedUrl: presignedUrl, contentType: "text/csv")
    }

    private func uploadSingleFile(path: String, presignedUrl: String, contentType: String) async throws {
        guard let url = URL(string: presignedUrl) else { throw AppError.unexpected(underlying: nil) }
        let fileURL = URL(fileURLWithPath: path)
        let data = try Data(contentsOf: fileURL)
        var req = URLRequest(url: url)
        req.httpMethod = "PUT"
        req.setValue(contentType, forHTTPHeaderField: "Content-Type")
        let (_, response) = try await session.upload(for: req, from: data)
        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw AppError.server(code: http.statusCode, message: "\(contentType) upload failed")
        }
    }

    // MARK: - Cached presigned URL accessors

    private func cachedSubmitResponse() -> SiteBlockCreateResponse? {
        guard let json = ApiCache.shared.getSubmitSiteBlock(),
              let data = json.data(using: .utf8) else { return nil }
        return try? JSONDecoder.tau.decode(SiteBlockCreateResponse.self, from: data)
    }

    func presignedVideoUrls() -> [String] {
        guard let resp = cachedSubmitResponse(), let url = resp.videoPresignedUrl else { return [] }
        return [url]
    }

    func presignedGpxUrl() -> String? { cachedSubmitResponse()?.gpxPresignedUrl }
    func presignedCsvUrl() -> String? { cachedSubmitResponse()?.csvPresignedUrl }
}
